webtranslate/pages/login.py

```diff
@@ -0,0 +1,71 @@
+"""
+Login/logout
+"""
+from urllib.parse import quote
+from webtranslate.bottle import route, request, redirect
+from webtranslate.protect import protected, start_session, stop_session
+from webtranslate import utils, users
+
+def login_success(r):
+    if r is None or not r.startswith("/"):
+        r = "/userprofile"
+
+    redirect(r + "?message=" + quote("Login successful!"))
+
+@route('/login', method = 'GET')
+@protected(['login', '-', '-'])
+def login(userauth):
+    req_redirect = request.query.get("redirect")
+    req_login = request.query.get('login')
+
+    if userauth.is_auth:
+        login_success(req_redirect)
+    elif users.oauth_redirect is not None:
+        session, url = users.oauth_redirect(req_redirect, req_login)
+        start_session(session)
+        redirect(url)
+    elif users.get_authentication is not None:
+        return utils.template('login', userauth = userauth, req_login = req_login, req_redirect = req_redirect)
+    else:
+        utils.redirect("/")
+
+@route('/login', method = 'POST')
+@protected(['login', '-', '-'])
+def login(userauth):
+    if users.get_authentication is None:
+        utils.redirect("/")
+        return
+
+    req_redirect = request.forms.get("redirect")
+    req_login = request.forms.get('login')
+    req_password = request.forms.get('password')
+
+    if req_login is not None and req_password is not None:
+        userauth = users.get_authentication(req_login, req_password)
+        if userauth.is_auth:
+            start_session(userauth)
+            login_success(req_redirect)
+            return
+
+    return utils.template('login', userauth = userauth, req_login = req_login, req_redirect = req_redirect, message = 'Try harder!', message_class = 'error')
+
+@route("/oauth2", method = 'GET')
+@protected(['oauth2', '-', '-'])
+def oauth(userauth):
+    if users.oauth_callback is None:
```

Comment:
This code makes a lot of use of `is None` and `is not None`, which is rather unusual for Python code to do. You only really use this in extreme cases where the difference between empty (`""`) and `None` is important (which rarely is. That heavily increases readability throughout the code. As I am pretty sure both `oauth_callback` of `""` and `None` should just follow this flow.